NiKomLib/Terminal.c

```diff
@@ -156,3 +156,220 @@ void __saveds __asm LIBStripAnsiSequences(register __a0 char *ansistr, register
                 tempstr=NULL;
         }
 }
+
+/* Lookup table for length of UTF-8 characters. */
+static const char utf8_extra[64] = {
+  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+  2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+  3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5
+};
+
+/* Name: ConvMBChrsToAmiga
+*  Parameters: a0 - Pointer to result string. Must be at least as big
+*                   as source string.
+*              a1 - Pointer to string that should be converted.
+*              d0 - Length of string to convert. If 0, convert until
+*                   nul.
+*              d1 - Source character set, defined in NiKomLib.h
+*
+*  Return value: Length of new string, or negative if more bytes are
+*                needed to perform conversion.
+*
+*  Description: Converts source string to ISO-8859-1 from the
+*               specified character set. Characters not present in
+*               ISO-8859-1 are replaced by '?'. For CHRS_CP437 and
+*               CHRS_MAC8 only characters greater than 128 are
+*               converted. For CHRS_SIS7 only characters greater than
+*               32 are converted.  Character greater than 128 in
+*               SIS-strings are interpreted as ISO-8859-1.
+*               CHRS_LATIN1 is not converted.
+*
+*               Note that the destination string is not nul terminated
+*               even if len is passed as zero. That must be done by
+*               the caller if needed.
+*/
+
+int __saveds __asm LIBConvMBChrsToAmiga(register __a0 char *dst,
+                                        register __a1 char *src,
+                                        register __d0 int len,
+                                        register __d1 int chrs,
+                                        register __a6 struct NiKomBase *NiKomBase) {
+  int si, di;
+
+  if(!len) {
+    len = INT_MAX;
+  }
+  for(si=0,di=0; src[si] && si<len; si++,di++) {
+    switch(chrs) {
+    case CHRS_CP437:
+      if(src[si] >= 128) {
+        dst[di] = NiKomBase->IbmToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_SIS7:
+      if(src[si] >= 32) {
+        dst[di] = NiKomBase->SF7ToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_MAC:
+      if(src[si] >= 128) {
+        dst[di] = NiKomBase->MacToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_CP850:
+      dst[di] = convnokludge(src[si]);
+      continue;
+    case CHRS_UTF8:
```

Comment:
Wouldn't it be both be more readable and performant if you created one function for each conversion and let these functions convert the entire string? Like you already have done for the UTF8  case with ParseUTF8().

This way the switch() to select the conversion won't be in the inner loop of the conversion, but just used to select the right conversion function once.